File: benchmarks/cross-language/zig/ping_pong.zig
// Zig ping-pong benchmark using std.Thread and Mutex
const std = @import("std");
const Thread = std.Thread;
const Mutex = std.Thread.Mutex;
const Condition = std.Thread.Condition;
const print = std.debug.print;

fn getMessages() usize {
    if (std.posix.getenv("BENCHMARK_MESSAGES")) |val| {
        return std.fmt.parseInt(usize, val, 10) catch 100_000;
    }
    return 100_000;
}

var MESSAGES: usize = 0;

const Channel = struct {
    mutex: Mutex = .{},
    cond: Condition = .{},
    ready: bool = false,
    value: i32 = 0,
};

var chan_a: Channel = .{};
var chan_b: Channel = .{};

fn pingThread(_: *anyopaque) void {
    var i: i32 = 0;
    while (i < MESSAGES) : (i += 1) {
        // Send to A
        chan_a.mutex.lock();
        chan_a.value = i;
        chan_a.ready = true;
        chan_a.cond.signal();
        chan_a.mutex.unlock();

        // Wait for B
        chan_b.mutex.lock();
        while (!chan_b.ready) {
            chan_b.cond.wait(&chan_b.mutex);
        }
        const received = chan_b.value;
        chan_b.ready = false;
        chan_b.mutex.unlock();

        // Validate received value matches what was sent
        if (received != i) {
            print("Ping validation error: expected {}, got {}\n", .{ i, received });
        }
    }
}

fn pongThread(_: *anyopaque) void {
    var expected: i32 = 0;
    while (expected < MESSAGES) : (expected += 1) {
        // Wait for A
        chan_a.mutex.lock();
        while (!chan_a.ready) {
            chan_a.cond.wait(&chan_a.mutex);
        }
        const received = chan_a.value;
        chan_a.ready = false;
        chan_a.mutex.unlock();

        // Validate received value matches expected sequence
        if (received != expected) {
            print("Pong validation error: expected {}, got {}\n", .{ expected, received });
        }

        // Send back the EXACT value received to B
        chan_b.mutex.lock();
        chan_b.value = received;
        chan_b.ready = true;
        chan_b.cond.signal();
        chan_b.mutex.unlock();
    }
}

fn getTimeNs() u64 {
    const ts = std.posix.clock_gettime(std.posix.CLOCK.MONOTONIC) catch return 0;
    return @as(u64, @intCast(ts.sec)) * 1_000_000_000 + @as(u64, @intCast(ts.nsec));
}

pub fn main() !void {
    MESSAGES = getMessages();

    print("=== Zig Ping-Pong Benchmark ===\n", .{});
    print("Messages: {}\n", .{MESSAGES});
    print("Using std.Thread with Mutex and Condition\n\n", .{});

    const start = getTimeNs();

    const thread1 = try Thread.spawn(.{}, pingThread, .{@as(*anyopaque, undefined)});
    const thread2 = try Thread.spawn(.{}, pongThread, .{@as(*anyopaque, undefined)});

    thread1.join();
    thread2.join();

    const end = getTimeNs();
    const elapsed_ns = end - start;
    const elapsed_sec = @as(f64, @floatFromInt(elapsed_ns)) / 1e9;
    const ns_per_msg = @as(f64, @floatFromInt(elapsed_ns)) / @as(f64, @floatFromInt(MESSAGES));
    const throughput = @as(f64, @floatFromInt(MESSAGES)) / elapsed_sec / 1e6;

    print("ns/msg:         {d:.2}\n", .{ns_per_msg});
    print("Throughput:     {d:.2} M msg/sec\n", .{throughput});
}

File: benchmarks/cross-language/c/ping_pong.c
// Pure C ping-pong benchmark with pthreads
// Baseline comparison showing raw pthread performance without Aether optimizations
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <stdint.h>
#include <unistd.h>
#include <time.h>

// Configurable via BENCHMARK_MESSAGES env var, default 100K for "low" preset
static int MESSAGES = 100000;

typedef struct {
    pthread_mutex_t mutex;
    pthread_cond_t cond;
    int ready;
    int value;
} Channel;

Channel *chan_a, *chan_b;

void* ping_thread(void* arg) {
    for (int i = 0; i < MESSAGES; i++) {
        // Send to A
        pthread_mutex_lock(&chan_a->mutex);
        chan_a->value = i;
        chan_a->ready = 1;
        pthread_cond_signal(&chan_a->cond);
        pthread_mutex_unlock(&chan_a->mutex);

        // Wait for B
        pthread_mutex_lock(&chan_b->mutex);
        while (!chan_b->ready) {
            pthread_cond_wait(&chan_b->cond, &chan_b->mutex);
        }
        int received_value = chan_b->value;
        chan_b->ready = 0;
        pthread_mutex_unlock(&chan_b->mutex);

        // VALIDATE: Must receive echo of what we sent
        if (received_value != i) {
            fprintf(stderr, "ERROR: Ping sent %d but got back %d\n", i, received_value);
        }
    }
    return NULL;
}

void* pong_thread(void* arg) {
    for (int i = 0; i < MESSAGES; i++) {
        // Wait for A
        pthread_mutex_lock(&chan_a->mutex);
        while (!chan_a->ready) {
            pthread_cond_wait(&chan_a->cond, &chan_a->mutex);
        }
        int received_value = chan_a->value;
        chan_a->ready = 0;
        pthread_mutex_unlock(&chan_a->mutex);

        // VALIDATE: Must receive expected sequence
        if (received_value != i) {
            fprintf(stderr, "ERROR: Pong expected %d but got %d\n", i, received_value);
        }

        // Send to B - echo back what we received
        pthread_mutex_lock(&chan_b->mutex);
        chan_b->value = received_value;
        chan_b->ready = 1;
        pthread_cond_signal(&chan_b->cond);
        pthread_mutex_unlock(&chan_b->mutex);
    }
    return NULL;
}

static inline uint64_t get_time_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ULL + ts.tv_nsec;
}

int main() {
    // Read message count from environment
    const char* env = getenv("BENCHMARK_MESSAGES");
    if (env) MESSAGES = atoi(env);

    chan_a = calloc(1, sizeof(Channel));
    chan_b = calloc(1, sizeof(Channel));

    pthread_mutex_init(&chan_a->mutex, NULL);
    pthread_cond_init(&chan_a->cond, NULL);
    pthread_mutex_init(&chan_b->mutex, NULL);
    pthread_cond_init(&chan_b->cond, NULL);

    printf("=== C (pthread) Ping-Pong Benchmark ===\n");
    printf("Messages: %d\n", MESSAGES);
    printf("Using pthread mutexes and condition variables\n\n");

    uint64_t start = get_time_ns();

    pthread_t t1, t2;
    pthread_create(&t1, NULL, ping_thread, NULL);
    pthread_create(&t2, NULL, pong_thread, NULL);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    uint64_t end = get_time_ns();
    double elapsed_sec = (double)(end - start) / 1e9;
    double throughput = MESSAGES / elapsed_sec;
    double ns_per_msg = elapsed_sec * 1e9 / MESSAGES;

    printf("ns/msg:         %.2f\n", ns_per_msg);
    printf("Throughput:     %.2f M msg/sec\n", throughput / 1e6);

    pthread_mutex_destroy(&chan_a->mutex);
    pthread_cond_destroy(&chan_a->cond);
    pthread_mutex_destroy(&chan_b->mutex);
    pthread_cond_destroy(&chan_b->cond);
    free(chan_a);
    free(chan_b);

    return 0;
}
